### ISSUE TITLE
Create Analytics tokens refresh job

### ISSUE THREAD

**Issue by victoralfaro-dotcms** [dotCMS/core#23245]
In order to keep DotCMS up with the Access Token generation to access the analytics stack we need a scheduled job.
This job will be in charge of:
- Running every now and then (frequency TBD) to detect when an Analytics Access Token is near to expire
- When the token is within in a configurable window of time, the job will try to refresh it in s separate thread
- During this window of time any other requests can consume the not-yet-expired token in order to avoid any blocking situation
- Retry a TBD number of times when the access token request fails
- Detect when Analytics stack is unrecoverable and set DotCMS in a state that it can notify that access tokens can no longer be refreshed

Flow for Access Token Request

* Is Auth in an unrecoverable error state, e.g. invalid secure secret, missing details
     * Return exception, need to direct to configuration page to update security.
* Is token in 1 min (configurable) window of expiry and update request thread (should not need a dedicated thread for this 
   just a background job in executor but should not run in parallel) not already running
	* yes
	  *  spawn thread to update token
			* On success
             	               * Update cached token
			* On failure
			       * If temporary error, e.g. network connection error, recoverable retry max 2 (delay ?) ?
                                * If a  permanent error, e.g. secret is wrong report  error notification exception, set status, do not keep retrying block further requests.
          * return old token immediately.
     * No Expired
		* spawn thread to update token if thread not already running
		* block for new token ( timeout auth exception, timeout must take into account any retry and delay, and preferably not fully block entire system, maybe add spinner until this timeout )
     * No
		* return existing token from cache no block.

**Comment by victoralfaro-dotcms** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/Falcon-6176e9d907eb0000106ee8e1/issues/dotCMS/core/23245?planning-poker) with Zenhub @dsilvam @freddyucv @oidacra

**Comment by dsilvam** [comment]
PR: https://github.com/dotCMS/core/pull/23494

**Comment by victoralfaro-dotcms** [comment]
Notes to QA:
- Start the analytics infrastructure by running from folder `<core>/cicd/docker` the following command:
```
docker-compose -f analytics-compose.yml up
```
- Wait for about 70 seconds for it to be loaded
- Add property to DotCMS configuration, probably by env-var, for `analytics.idp.url` property with the value: http://localhost:61111/realms/dotcms/protocol/openid-connect/token and another for `analytics.app.config.url` with the value: http://localhost:8088/c/customer1/cluster1/keys
- Create a Analytics Application with a test clientId and secret: `analytics-customer-customer1`/`testsecret`
- The job runs every minute so wait for at least a minute to verify that the logs has something like:
```
Starting access token renew thread for clientIds 
```

**Comment by bryanboza** [comment]
This is not working in the release branch, we have some errors here.

- We are printing the access token in the log. By security we can't do that 
```
17:50:00.106  ERROR analytics.AnalyticsAPIImpl - Error requesting ACCESS_TOKEN with clientId analytics-customer-customer1 from IDP server http://localhost:61111/realms/dotcms/protocol/openid-connect/token (response: Response{response=AccessToken{accessToken=eyJhbGciOiJSUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICI2WlNESmFQWmxuWmV4VWgwZkhRaE5QNVJZTHBGc2JCNXFJaUhTWHYtU3JjIn0.eyJleHAiOjE2NzUxMjI5MDAsImlhdCI6MTY3NTEyMjYwMCwianRpIjoiODYyZmViNDctODMxNi00ZGI1LTkyMzItNjQxNGE1MTYzZjk4IiwiaXNzIjoiaHR0cDovL2xvY2FsaG9zdDo2MTExMS9yZWFsbXMvZG90Y21zIiwiYXVkIjoiYXBpLWRvdGNtcy1hbmFseXRpY3MtYXVkaWVuY2UiLCJzdWIiOiI2NDQwYTY2ZS1iOWRhLTQ4MTgtYThlNS03YzQ0M2I2Y2U4NmEiLCJ0eXAiOiJCZWFyZXIiLCJhenAiOiJhbmFseXRpY3MtY3VzdG9tZXItY3VzdG9tZXIxIiwic2NvcGUiOiIiLCJodHRwczovL2RvdGNtcy5jb20vYW5hbHl0aWNzL3JvbGVzIjpbImN1c3RvbWVyIl0sImNsaWVudElkIjoiYW5hbHl0aWNzLWN1c3RvbWVyLWN1c3RvbWVyMSIsImNsaWVudEhvc3QiOiIxNzIuMjEuMC4xIiwiaHR0cHM6Ly9kb3RjbXMuY29tL2FuYWx5dGljcyI6eyJjdXN0b21lcklkIjoiY3VzdG9tZXIxIiwiY2x1c3RlcklkIjoiY2x1c3RlcjEifSwiY2xpZW50QWRkcmVzcyI6IjE3Mi4yMS4wLjEifQ.NrflY3MOYG4TbHKTFjjVJVzAMij7xiGhK-e_vwkQ97slIhEhtqQe-GLKTJvS5qxy7ZBrjly-HCHGq8fxwa_zhHmS5uQqBY4UXkxOxzLpA8zNS0qoAhgiyDGnVj6_ixoRCGm2xDbcF_FULei4KNON210Jz0EiRKF8Cj-fZulObVppec_7fHdZMtB8vQ-Qq6yDbXv-Lm4QzemE2Fxed6wMKwrdTc1b1moXDELO4l7eHw4LUCUmaY2ztaDrzQIVOr2DJsELQzDf5rDg1qP1eDI9IYGEFVgUc2m6TGkl8ly256C-1H2SBmgIncyRCWR8_DpbpiX7f5NmP1qdLW7p6E21dQ, tokenType=Bearer, expiresIn=300, refreshExpiresIn=0, scope=}, statusCode=200})
```

Also, We are unable to refresh the token when is expired.

**Comment by mbiuki** [comment]
This looks like you are sharing a jwt token in logs and it must be prohibited. Additionally, there are certain best practices to make sure jwt tokens are created securely, here are some recommendations:

1. Use strong cryptography: Use a secure algorithm such as HS256, HS384, or HS512 for signing tokens.
3. Keep the token size small: The token size should be kept as small as possible to reduce network overhead.
5. Use short expiration times: Tokens should have a short expiration time, such as 15 minutes, to reduce the risk of theft.
7. Store tokens securely: Tokens should be stored securely on the client-side, such as in HTTP-only and secure cookies.
9. Verify the signature: The signature of the token should be verified on the server to prevent tampering.
11. Avoid storing sensitive information in the token: Tokens should not contain sensitive information such as passwords, as they can be easily decoded.
13. Revoke tokens: Tokens should be revoked when the user logs out or when their session expires.
15. Use HTTPS: All communication involving JWT should be done over HTTPS to prevent eavesdropping.

**Comment by dsilvam** [comment]
Another PR: https://github.com/dotCMS/core/pull/23952

**Comment by bryanboza** [comment]
Fixed, tested on release-23.02 and now we are renovating the key successfully

```
12:35:00.105  INFO  analytics.AccessTokenRenewRunnable - ACCESS_TOKEN for clientId analytics-customer-customer1 has been successfully renewed
```

Tested following the provided information in the previous comments.
